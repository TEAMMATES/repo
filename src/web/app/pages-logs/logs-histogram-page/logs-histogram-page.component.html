<h1>Logs Histogram View</h1>
<p>This page provides a histogram view for error logs in the past {{ LOGS_RETENTION_PERIOD_IN_DAYS }} days.</p>
<button (click)="toLogsPage()" class="btn btn-info">Go to logs page</button>
<hr/>

<div *tmIsLoading="isLoading" class="card bg-form no-border">
  <div class="card-body">
    <div class="row form-group">
      <div class="col-5">
        <div class="col-12 mb-1">
          <label for="logs-from-datepicker" class="font-weight-bold">Search period from</label>
        </div>
        <div class="input-group">
          <div class="col-md-6 input-group">
            <input id="logs-from-datepicker" type="text" class="form-control" ngbDatepicker [minDate]="earliestSearchDate" [maxDate]="dateToday" [(ngModel)]="formModel.logsDateFrom" #logsFromDp="ngbDatepicker"/>
            <button class="btn btn-light" (click)="logsFromDp.toggle()" type="button">
              <i class="fas fa-calendar-alt"></i>
            </button>
          </div>
          <div id="logs-from-timepicker" class="col-md-4">
            <tm-time-picker [(time)]="formModel.logsTimeFrom"></tm-time-picker>
          </div>
        </div>
      </div>
      <div class="col-5">
        <div class="col-12 mb-1">
          <label for="logs-to-datepicker" class="font-weight-bold">Search period until</label>
        </div>
        <div class="input-group">
          <div class="col-md-6 input-group">
            <input id="logs-to-datepicker" type="text" class="form-control" ngbDatepicker [minDate]="earliestSearchDate" [maxDate]="dateToday" [(ngModel)]="formModel.logsDateTo" #logsToDp="ngbDatepicker"/>
            <button class="btn btn-light" (click)="logsToDp.toggle()" type="button">
              <i class="fas fa-calendar-alt"></i>
            </button>
          </div>
          <div id="logs-to-timepicker" class="col-md-4">
            <tm-time-picker [(time)]="formModel.logsTimeTo"></tm-time-picker>
          </div>
        </div>
      </div>
      <div class="col-1 mt-4">
        <button id="query-button" class="float-right btn btn-primary" (click)="searchForLogs()">Query</button>
      </div>
    </div>
  </div>
</div>

<div id="histogram" style="height: 400px; width: 100%;" *tmIsLoading="isSearching">
  <tm-logs-histogram [data]="searchResult" *ngIf="searchResult.length>0; else noDataMessage"></tm-logs-histogram>
</div>

<ng-template #noDataMessage>
  <div class="no-data-message">
    There are no error logs in this time period.
  </div>
</ng-template>
