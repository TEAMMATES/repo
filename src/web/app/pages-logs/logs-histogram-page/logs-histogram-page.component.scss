.no-border {
  border: 0;
}

.bg-form {
  background-color: #EAEFF5;
}
