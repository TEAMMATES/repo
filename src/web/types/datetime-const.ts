export enum HoursConst {
    ZERO_HOURS = 0,
    TWELVE_HOURS = 12,
    ONE_DAY_HOURS = 24,
    THREE_DAYS_HOURS = 72,
    ONE_WEEK_HOURS = 168,
}

export enum SecondsConst {
    ONE_MINUTE_SECONDS = 60,
}

export enum MinutesConst {
    ONE_HOURS_MINUTES = 60,
}

export enum MsConst {
    ONE_SECOND_MILLISECONDS = 1000,
    TEN_SECOND_MILLISECONDS = 10000,
    ONE_MINUTE_MILLISECONDS = 60000,
    TEN_MINUTE_MILLISECONDS = 600000,
    FIFTEEN_MINUTE_MILLISECONDS = 900000,
    ONE_HOUR_MILLISECONDS = 3600000,
    ONE_WEEK_MILLISECONDS = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

/**
 * The date format.
 */
export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

/**
 * The output format of the time picker.
 */
export interface TimeFormat {
    hour: number;
    minute: number;
}
