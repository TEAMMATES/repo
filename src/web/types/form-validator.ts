import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify URLs as Angular does not have a built-in URL validator.
   */
  URL_REGEX = ApiStringConst.URL_REGEX,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,
}
