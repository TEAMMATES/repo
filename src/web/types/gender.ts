/**
 * Represents the gender of a student.
 */
export enum Gender {
  MALE,
  FEMALE,
  OTHER,
}
export namespace Gender {

  export const values: () => string[] = (): string[] =>
      Object.keys(Gender).filter((key: string) => isNaN(key as any) && key !== 'values');

  export const enums: () => Gender[] = (): Gender[] => {
    const genders: Gender[] = [];
    genders.push(Gender.MALE);
    genders.push(Gender.FEMALE);
    genders.push(Gender.OTHER);
    return genders;
  };
}
