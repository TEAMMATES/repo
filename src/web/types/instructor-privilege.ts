import { InstructorPrivilege } from './api-output';

/**
 * The default instructor privilege.
 */
export const DEFAULT_INSTRUCTOR_PRIVILEGE: InstructorPrivilege = {
  canEditCourse: false,
  canDeleteCourse: false,
  canRestoreCourse: false,
  canModifySession: false,
  canModifyStudent: false,
  canSubmitSessionInSections: false,
  canModifyInstructor: false,
  canViewStudentInSections: false,
  canModifySessionCommentsInSections: false,
  canViewSessionInSections: false,
};
