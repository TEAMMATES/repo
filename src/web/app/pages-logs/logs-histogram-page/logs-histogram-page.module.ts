import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule, Routes } from '@angular/router';
import { NgbDatepickerModule } from '@ng-bootstrap/ng-bootstrap';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { LogsHistogramModule } from '../../components/logs-histogram/logs-histogram.module';
import { SessionEditFormModule } from '../../components/session-edit-form/session-edit-form.module';
import { LogsHistogramPageComponent } from './logs-histogram-page.component';

const routes: Routes = [
  {
    path: '',
    component: LogsHistogramPageComponent,
  },
];

/**
 * Module for logs histogram view page.
 */
@NgModule({
  imports: [
    CommonModule,
    RouterModule.forChild(routes),
    NgbDatepickerModule,
    FormsModule,
    LoadingSpinnerModule,
    SessionEditFormModule,
    LogsHistogramModule,
  ],
  declarations: [
    LogsHistogramPageComponent,
  ],
  exports: [
    LogsHistogramPageComponent,
  ],
})
export class LogsHistogramPageModule { }
