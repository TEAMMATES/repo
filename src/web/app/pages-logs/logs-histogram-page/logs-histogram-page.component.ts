import { Component, OnInit } from '@angular/core';
import moment from 'moment-timezone';
import { EMPTY, forkJoin, Observable } from 'rxjs';
import { concatMap, expand, finalize, map, reduce } from 'rxjs/operators';
import { LogService } from 'src/web/services/log.service';
import { StatusMessageService } from 'src/web/services/status-message.service';
import { LOCAL_DATE_TIME_FORMAT, TimeResolvingResult, TimezoneService } from 'src/web/services/timezone.service';
import { GeneralLogEntry, GeneralLogs } from 'src/web/types/api-output';
import { ApiConst } from '../../../types/api-const';
import { LogsHistogramDataModel } from '../../components/logs-histogram/logs-histogram-model';
import { DateFormat } from '../../components/session-edit-form/session-edit-form-model';
import { TimeFormat } from '../../components/session-edit-form/time-picker/time-picker.component';
import { ErrorMessageOutput } from '../../error-message-output';

/**
 * Model for searching of logs
 */
 interface SearchLogsFormModel {
  logsDateFrom: DateFormat;
  logsDateTo: DateFormat;
  logsTimeFrom: TimeFormat;
  logsTimeTo: TimeFormat;
}

/**
 * Query parameters for HTTP request
 */
 interface QueryParams {
  searchFrom: string;
  searchUntil: string;
  severity: string;
  nextPageToken?: string;
 }

@Component({
  selector: 'tm-logs-histogram-page',
  templateUrl: './logs-histogram-page.component.html',
  styleUrls: ['./logs-histogram-page.component.scss']
})
export class LogsHistogramPageComponent implements OnInit {

  readonly LOGS_RETENTION_PERIOD_IN_DAYS: number = ApiConst.LOGS_RETENTION_PERIOD;
  readonly LOGS_RETENTION_PERIOD_IN_MILLISECONDS: number = this.LOGS_RETENTION_PERIOD_IN_DAYS * 24 * 60 * 60 * 1000;
  
  formModel: SearchLogsFormModel = {
    logsDateFrom: { year: 0, month: 0, day: 0 },
    logsTimeFrom: { hour: 0, minute: 0 },
    logsDateTo: { year: 0, month: 0, day: 0 },
    logsTimeTo: { hour: 0, minute: 0 },
  }
  previousQueryParams: QueryParams = {
    searchFrom: '',
    searchUntil: '',
    severity: '',
  }
  dateToday: DateFormat = { year: 0, month: 0, day: 0 };
  earliestSearchDate: DateFormat = { year: 0, month: 0, day: 0 };
  nextPageToken: string | undefined = '';
  isLoading: boolean = false;
  isSearching: boolean = false;
  searchResult: LogsHistogramDataModel[] = [];

  constructor(private logService: LogService, 
    private timezoneService: TimezoneService,
    private statusMessageService: StatusMessageService) { }

  ngOnInit(): void {
    const today: Date = new Date();
    this.dateToday.year = today.getFullYear();
    this.dateToday.month = today.getMonth() + 1;
    this.dateToday.day = today.getDate();

    const earliestSearchDate: Date = new Date(Date.now() - this.LOGS_RETENTION_PERIOD_IN_MILLISECONDS);
    this.earliestSearchDate.year = earliestSearchDate.getFullYear();
    this.earliestSearchDate.month = earliestSearchDate.getMonth() + 1;
    this.earliestSearchDate.day = earliestSearchDate.getDate();

    const fromDate: Date = new Date();
    fromDate.setDate(today.getDate() - 1);

    this.formModel.logsDateFrom = {
      year: fromDate.getFullYear(),
      month: fromDate.getMonth() + 1,
      day: fromDate.getDate(),
    };
    this.formModel.logsDateTo = { ...this.dateToday };
    this.formModel.logsTimeFrom = { hour: 23, minute: 59 };
    this.formModel.logsTimeTo = { hour: 23, minute: 59 };
  }

  searchForLogs(): void {
    this.isSearching = true;
    this.searchResult = [];
    const localDateTime: Observable<number>[] = [
      this.resolveLocalDateTime(this.formModel.logsDateFrom, this.formModel.logsTimeFrom, 'Search period from'),
      this.resolveLocalDateTime(this.formModel.logsDateTo, this.formModel.logsTimeTo, 'Search period until'),
    ];
    forkJoin(localDateTime)
      .pipe(
        concatMap(([timestampFrom, timestampUntil]: number[]) => {
          this.previousQueryParams = {
            searchFrom: timestampFrom.toString(),
            searchUntil: timestampUntil.toString(),
            severity: 'ERROR', 
          }
          return this.logService.searchLogs(this.previousQueryParams);
        }),
        finalize(() => {
          this.isSearching = false;
          // this.hasResult = true;
        }))
      .pipe(
        expand((logs: GeneralLogs) => {
          if (logs.nextPageToken !== undefined) {
            this.previousQueryParams.nextPageToken = logs.nextPageToken;
            return this.logService.searchLogs(this.previousQueryParams);
          }

          return EMPTY;
        }),
        reduce((acc, res) => acc.concat(res.logEntries), [] as GeneralLogEntry[])
      )
      .subscribe((logResults: GeneralLogEntry[]) => this.processLogs(logResults),
        (e: ErrorMessageOutput) => this.statusMessageService.showErrorToast(e.error.message));
  }

  private processLogs(logs: GeneralLogEntry[]): void {
    const map: Map<string, number> = logs.reduce((acc, log) =>
      acc.set(JSON.stringify(log.sourceLocation), (acc.get(JSON.stringify(log.sourceLocation)) || 0) + 1),
      new Map<string, number>());
    map.forEach((value: number, key: string) => {
      this.searchResult.push({ sourceLocation: JSON.parse(key), numberOfTimes: value });
    });
  }

  private resolveLocalDateTime(date: DateFormat, time: TimeFormat, fieldName: string): Observable<number> {
    const inst: any = moment();
    inst.set('year', date.year);
    inst.set('month', date.month - 1);
    inst.set('date', date.day);
    inst.set('hour', time.hour);
    inst.set('minute', time.minute);
    const localDateTime: string = inst.format(LOCAL_DATE_TIME_FORMAT);

    return this.timezoneService.getResolvedTimestamp(localDateTime, this.timezoneService.guessTimezone(), fieldName)
        .pipe(map((result: TimeResolvingResult) => result.timestamp));
  }
}
