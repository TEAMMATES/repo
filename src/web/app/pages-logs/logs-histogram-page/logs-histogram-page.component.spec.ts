import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';
import { LogService } from '../../../services/log.service';
import { TimezoneService } from '../../../services/timezone.service';
import { GeneralLogEntry } from '../../../types/api-output';
import { LogsHistogramPageComponent } from './logs-histogram-page.component';
import { LogsHistogramPageModule } from './logs-histogram-page.module';
import Spy = jasmine.Spy;

describe('LogsHistogramPageComponent', () => {
  let component: LogsHistogramPageComponent;
  let fixture: ComponentFixture<LogsHistogramPageComponent>;
  let logService: LogService;
  let timezoneService: TimezoneService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [NgbModule, LogsHistogramPageModule, HttpClientTestingModule, RouterTestingModule],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(LogsHistogramPageComponent);
    component = fixture.componentInstance;
    logService = TestBed.inject(LogService);
    timezoneService = TestBed.inject(TimezoneService);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when page is still loading', () => {
    component.isLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when searching for details in search form', () => {
    component.formModel = {
      logsDateFrom: { year: 2021, month: 6, day: 1 },
      logsTimeFrom: { hour: 23, minute: 59 },
      logsDateTo: { year: 2021, month: 6, day: 2 },
      logsTimeTo: { hour: 23, minute: 59 },
    };
    component.isLoading = false;
    component.isSearching = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should search for all logs when search button is clicked', () => {
    const logSpy: Spy = spyOn(logService, 'searchLogs').and
        .returnValues(of({ logEntries: [], nextPageToken: 'token' }), of({ logEntries: [] }));
    const timeSpy: Spy = spyOn(timezoneService, 'getResolvedTimestamp').and
        .returnValue(of({ timestamp: 0, message: '' }));

    component.isLoading = false;
    component.isSearching = false;
    component.formModel = {
      logsDateFrom: { year: 2021, month: 6, day: 1 },
      logsTimeFrom: { hour: 23, minute: 59 },
      logsDateTo: { year: 2021, month: 6, day: 2 },
      logsTimeTo: { hour: 23, minute: 59 },
    };
    fixture.detectChanges();

    fixture.debugElement.nativeElement.querySelector('#query-button').click();

    expect(timeSpy).toHaveBeenCalledTimes(2);
    expect(logSpy).toHaveBeenCalledTimes(2);
    expect(logSpy.calls.mostRecent().args).toEqual([{
      searchFrom: '0',
      searchUntil: '0',
      severity: 'ERROR',
      nextPageToken: 'token',
    }]);
  });

  it('should sort logs based on source location', () => {
    const testLog1: GeneralLogEntry = {
      logName: 'stderr',
      severity: 'ERROR',
      trace: 'testTrace1',
      sourceLocation: {
        file: 'file1',
        line: 10,
        function: 'function1',
      },
      timestamp: 1549095330000,
      message: 'message',
    };
    const testLog2: GeneralLogEntry = {
      logName: 'stderr',
      severity: 'ERROR',
      trace: 'testTrace2',
      sourceLocation: {
        file: 'file2',
        line: 10,
        function: 'function1',
      },
      timestamp: 1549095330000,
      message: 'message',
    };
    const testLog3: GeneralLogEntry = {
      logName: 'stderr',
      severity: 'ERROR',
      trace: 'testTrace3',
      sourceLocation: {
        file: 'file2',
        line: 10,
        function: 'function1',
      },
      timestamp: 1549095330000,
      message: 'message',
    };
    spyOn(logService, 'searchLogs').and
      .returnValue(of({ logEntries: [testLog1, testLog2, testLog3] }));
    spyOn(timezoneService, 'getResolvedTimestamp').and.returnValue(of({ timestamp: 0, message: '' }));

    component.searchForLogs();

    expect(component.searchResult.length).toEqual(2);
  });
});
