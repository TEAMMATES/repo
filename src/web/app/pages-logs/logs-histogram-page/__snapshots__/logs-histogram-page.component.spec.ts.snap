// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogsHistogramPageComponent should snap when page is still loading 1`] = `
<tm-logs-histogram-page
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  dateToday={[Function Object]}
  earliestSearchDate={[Function Object]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading={[Function Boolean]}
  isSearching="false"
  logService={[Function LogService]}
  nextPageToken=""
  previousQueryParams={[Function Object]}
  route={[Function Router]}
  searchResult={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Logs Histogram View
  </h1><p>
    This page provides a histogram view for error logs in the past 30 days.
  </p><button
    class="btn btn-info"
  >
    Go to logs page
  </button><hr /><tm-loading-spinner>
    <div
      class="loading-container"
    >
      <div
        class="row spinner-border text-primary"
        role="status"
      />
      <div
        class="row text-secondary"
      >
        Loading...
      </div>
    </div>
  </tm-loading-spinner><div
    id="histogram"
    style="height: 400px; width: 100%;"
  >
    
    <div
      class="no-data-message"
    >
       There are no error logs in this time period. 
    </div>
  </div>
</tm-logs-histogram-page>
`;

exports[`LogsHistogramPageComponent should snap when searching for details in search form 1`] = `
<tm-logs-histogram-page
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  dateToday={[Function Object]}
  earliestSearchDate={[Function Object]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading="false"
  isSearching={[Function Boolean]}
  logService={[Function LogService]}
  nextPageToken=""
  previousQueryParams={[Function Object]}
  route={[Function Router]}
  searchResult={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Logs Histogram View
  </h1><p>
    This page provides a histogram view for error logs in the past 30 days.
  </p><button
    class="btn btn-info"
  >
    Go to logs page
  </button><hr /><div
    class="card bg-form no-border"
  >
    <div
      class="card-body"
    >
      <div
        class="row form-group"
      >
        <div
          class="col-5"
        >
          <div
            class="col-12 mb-1"
          >
            <label
              class="font-weight-bold"
              for="logs-from-datepicker"
            >
              Search period from
            </label>
          </div>
          <div
            class="input-group"
          >
            <div
              class="col-md-6 input-group"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                id="logs-from-datepicker"
                ngbdatepicker=""
                type="text"
              />
              <button
                class="btn btn-light"
                type="button"
              >
                <i
                  class="fas fa-calendar-alt"
                />
              </button>
            </div>
            <div
              class="col-md-4"
              id="logs-from-timepicker"
            >
              <tm-time-picker>
                <select
                  class="form-control ng-untouched ng-pristine ng-valid"
                >
                  
                  <option
                    value="1: Object"
                  >
                     01:00H 
                  </option>
                  <option
                    value="2: Object"
                  >
                     02:00H 
                  </option>
                  <option
                    value="3: Object"
                  >
                     03:00H 
                  </option>
                  <option
                    value="4: Object"
                  >
                     04:00H 
                  </option>
                  <option
                    value="5: Object"
                  >
                     05:00H 
                  </option>
                  <option
                    value="6: Object"
                  >
                     06:00H 
                  </option>
                  <option
                    value="7: Object"
                  >
                     07:00H 
                  </option>
                  <option
                    value="8: Object"
                  >
                     08:00H 
                  </option>
                  <option
                    value="9: Object"
                  >
                     09:00H 
                  </option>
                  <option
                    value="10: Object"
                  >
                     10:00H 
                  </option>
                  <option
                    value="11: Object"
                  >
                     11:00H 
                  </option>
                  <option
                    value="12: Object"
                  >
                     12:00H 
                  </option>
                  <option
                    value="13: Object"
                  >
                     13:00H 
                  </option>
                  <option
                    value="14: Object"
                  >
                     14:00H 
                  </option>
                  <option
                    value="15: Object"
                  >
                     15:00H 
                  </option>
                  <option
                    value="16: Object"
                  >
                     16:00H 
                  </option>
                  <option
                    value="17: Object"
                  >
                     17:00H 
                  </option>
                  <option
                    value="18: Object"
                  >
                     18:00H 
                  </option>
                  <option
                    value="19: Object"
                  >
                     19:00H 
                  </option>
                  <option
                    value="20: Object"
                  >
                     20:00H 
                  </option>
                  <option
                    value="21: Object"
                  >
                     21:00H 
                  </option>
                  <option
                    value="22: Object"
                  >
                     22:00H 
                  </option>
                  <option
                    value="23: Object"
                  >
                     23:00H 
                  </option>
                  <option
                    value="0: Object"
                  >
                    23:59H
                  </option>
                  
                </select>
              </tm-time-picker>
            </div>
          </div>
        </div>
        <div
          class="col-5"
        >
          <div
            class="col-12 mb-1"
          >
            <label
              class="font-weight-bold"
              for="logs-to-datepicker"
            >
              Search period until
            </label>
          </div>
          <div
            class="input-group"
          >
            <div
              class="col-md-6 input-group"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                id="logs-to-datepicker"
                ngbdatepicker=""
                type="text"
              />
              <button
                class="btn btn-light"
                type="button"
              >
                <i
                  class="fas fa-calendar-alt"
                />
              </button>
            </div>
            <div
              class="col-md-4"
              id="logs-to-timepicker"
            >
              <tm-time-picker>
                <select
                  class="form-control ng-untouched ng-pristine ng-valid"
                >
                  
                  <option
                    value="1: Object"
                  >
                     01:00H 
                  </option>
                  <option
                    value="2: Object"
                  >
                     02:00H 
                  </option>
                  <option
                    value="3: Object"
                  >
                     03:00H 
                  </option>
                  <option
                    value="4: Object"
                  >
                     04:00H 
                  </option>
                  <option
                    value="5: Object"
                  >
                     05:00H 
                  </option>
                  <option
                    value="6: Object"
                  >
                     06:00H 
                  </option>
                  <option
                    value="7: Object"
                  >
                     07:00H 
                  </option>
                  <option
                    value="8: Object"
                  >
                     08:00H 
                  </option>
                  <option
                    value="9: Object"
                  >
                     09:00H 
                  </option>
                  <option
                    value="10: Object"
                  >
                     10:00H 
                  </option>
                  <option
                    value="11: Object"
                  >
                     11:00H 
                  </option>
                  <option
                    value="12: Object"
                  >
                     12:00H 
                  </option>
                  <option
                    value="13: Object"
                  >
                     13:00H 
                  </option>
                  <option
                    value="14: Object"
                  >
                     14:00H 
                  </option>
                  <option
                    value="15: Object"
                  >
                     15:00H 
                  </option>
                  <option
                    value="16: Object"
                  >
                     16:00H 
                  </option>
                  <option
                    value="17: Object"
                  >
                     17:00H 
                  </option>
                  <option
                    value="18: Object"
                  >
                     18:00H 
                  </option>
                  <option
                    value="19: Object"
                  >
                     19:00H 
                  </option>
                  <option
                    value="20: Object"
                  >
                     20:00H 
                  </option>
                  <option
                    value="21: Object"
                  >
                     21:00H 
                  </option>
                  <option
                    value="22: Object"
                  >
                     22:00H 
                  </option>
                  <option
                    value="23: Object"
                  >
                     23:00H 
                  </option>
                  <option
                    value="0: Object"
                  >
                    23:59H
                  </option>
                  
                </select>
              </tm-time-picker>
            </div>
          </div>
        </div>
        <div
          class="col-1 mt-4"
        >
          <button
            class="float-right btn btn-primary"
            id="query-button"
          >
            Query
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-spinner>
    <div
      class="loading-container"
    >
      <div
        class="row spinner-border text-primary"
        role="status"
      />
      <div
        class="row text-secondary"
      >
        Loading...
      </div>
    </div>
  </tm-loading-spinner>
</tm-logs-histogram-page>
`;

exports[`LogsHistogramPageComponent should snap with default fields 1`] = `
<tm-logs-histogram-page
  LOGS_RETENTION_PERIOD_IN_DAYS={[Function Number]}
  LOGS_RETENTION_PERIOD_IN_MILLISECONDS={[Function Number]}
  dateToday={[Function Object]}
  earliestSearchDate={[Function Object]}
  formModel={[Function Object]}
  hasResult="false"
  isLoading="false"
  isSearching="false"
  logService={[Function LogService]}
  nextPageToken=""
  previousQueryParams={[Function Object]}
  route={[Function Router]}
  searchResult={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Logs Histogram View
  </h1><p>
    This page provides a histogram view for error logs in the past 30 days.
  </p><button
    class="btn btn-info"
  >
    Go to logs page
  </button><hr /><div
    class="card bg-form no-border"
  >
    <div
      class="card-body"
    >
      <div
        class="row form-group"
      >
        <div
          class="col-5"
        >
          <div
            class="col-12 mb-1"
          >
            <label
              class="font-weight-bold"
              for="logs-from-datepicker"
            >
              Search period from
            </label>
          </div>
          <div
            class="input-group"
          >
            <div
              class="col-md-6 input-group"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                id="logs-from-datepicker"
                ngbdatepicker=""
                type="text"
              />
              <button
                class="btn btn-light"
                type="button"
              >
                <i
                  class="fas fa-calendar-alt"
                />
              </button>
            </div>
            <div
              class="col-md-4"
              id="logs-from-timepicker"
            >
              <tm-time-picker>
                <select
                  class="form-control ng-untouched ng-pristine ng-valid"
                >
                  
                  <option
                    value="1: Object"
                  >
                     01:00H 
                  </option>
                  <option
                    value="2: Object"
                  >
                     02:00H 
                  </option>
                  <option
                    value="3: Object"
                  >
                     03:00H 
                  </option>
                  <option
                    value="4: Object"
                  >
                     04:00H 
                  </option>
                  <option
                    value="5: Object"
                  >
                     05:00H 
                  </option>
                  <option
                    value="6: Object"
                  >
                     06:00H 
                  </option>
                  <option
                    value="7: Object"
                  >
                     07:00H 
                  </option>
                  <option
                    value="8: Object"
                  >
                     08:00H 
                  </option>
                  <option
                    value="9: Object"
                  >
                     09:00H 
                  </option>
                  <option
                    value="10: Object"
                  >
                     10:00H 
                  </option>
                  <option
                    value="11: Object"
                  >
                     11:00H 
                  </option>
                  <option
                    value="12: Object"
                  >
                     12:00H 
                  </option>
                  <option
                    value="13: Object"
                  >
                     13:00H 
                  </option>
                  <option
                    value="14: Object"
                  >
                     14:00H 
                  </option>
                  <option
                    value="15: Object"
                  >
                     15:00H 
                  </option>
                  <option
                    value="16: Object"
                  >
                     16:00H 
                  </option>
                  <option
                    value="17: Object"
                  >
                     17:00H 
                  </option>
                  <option
                    value="18: Object"
                  >
                     18:00H 
                  </option>
                  <option
                    value="19: Object"
                  >
                     19:00H 
                  </option>
                  <option
                    value="20: Object"
                  >
                     20:00H 
                  </option>
                  <option
                    value="21: Object"
                  >
                     21:00H 
                  </option>
                  <option
                    value="22: Object"
                  >
                     22:00H 
                  </option>
                  <option
                    value="23: Object"
                  >
                     23:00H 
                  </option>
                  <option
                    value="0: Object"
                  >
                    23:59H
                  </option>
                  
                </select>
              </tm-time-picker>
            </div>
          </div>
        </div>
        <div
          class="col-5"
        >
          <div
            class="col-12 mb-1"
          >
            <label
              class="font-weight-bold"
              for="logs-to-datepicker"
            >
              Search period until
            </label>
          </div>
          <div
            class="input-group"
          >
            <div
              class="col-md-6 input-group"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                id="logs-to-datepicker"
                ngbdatepicker=""
                type="text"
              />
              <button
                class="btn btn-light"
                type="button"
              >
                <i
                  class="fas fa-calendar-alt"
                />
              </button>
            </div>
            <div
              class="col-md-4"
              id="logs-to-timepicker"
            >
              <tm-time-picker>
                <select
                  class="form-control ng-untouched ng-pristine ng-valid"
                >
                  
                  <option
                    value="1: Object"
                  >
                     01:00H 
                  </option>
                  <option
                    value="2: Object"
                  >
                     02:00H 
                  </option>
                  <option
                    value="3: Object"
                  >
                     03:00H 
                  </option>
                  <option
                    value="4: Object"
                  >
                     04:00H 
                  </option>
                  <option
                    value="5: Object"
                  >
                     05:00H 
                  </option>
                  <option
                    value="6: Object"
                  >
                     06:00H 
                  </option>
                  <option
                    value="7: Object"
                  >
                     07:00H 
                  </option>
                  <option
                    value="8: Object"
                  >
                     08:00H 
                  </option>
                  <option
                    value="9: Object"
                  >
                     09:00H 
                  </option>
                  <option
                    value="10: Object"
                  >
                     10:00H 
                  </option>
                  <option
                    value="11: Object"
                  >
                     11:00H 
                  </option>
                  <option
                    value="12: Object"
                  >
                     12:00H 
                  </option>
                  <option
                    value="13: Object"
                  >
                     13:00H 
                  </option>
                  <option
                    value="14: Object"
                  >
                     14:00H 
                  </option>
                  <option
                    value="15: Object"
                  >
                     15:00H 
                  </option>
                  <option
                    value="16: Object"
                  >
                     16:00H 
                  </option>
                  <option
                    value="17: Object"
                  >
                     17:00H 
                  </option>
                  <option
                    value="18: Object"
                  >
                     18:00H 
                  </option>
                  <option
                    value="19: Object"
                  >
                     19:00H 
                  </option>
                  <option
                    value="20: Object"
                  >
                     20:00H 
                  </option>
                  <option
                    value="21: Object"
                  >
                     21:00H 
                  </option>
                  <option
                    value="22: Object"
                  >
                     22:00H 
                  </option>
                  <option
                    value="23: Object"
                  >
                     23:00H 
                  </option>
                  <option
                    value="0: Object"
                  >
                    23:59H
                  </option>
                  
                </select>
              </tm-time-picker>
            </div>
          </div>
        </div>
        <div
          class="col-1 mt-4"
        >
          <button
            class="float-right btn btn-primary"
            id="query-button"
          >
            Query
          </button>
        </div>
      </div>
    </div>
  </div><div
    id="histogram"
    style="height: 400px; width: 100%;"
  >
    
    <div
      class="no-data-message"
    >
       There are no error logs in this time period. 
    </div>
  </div>
</tm-logs-histogram-page>
`;
